### PR TITLE
filewatcher: Fix flaky tests

### PR DESCRIPTION
**Pull request by fishy**
When overwriting the file with new content for test, the overwriting was
done incorrectly, via overwriting in place, and can sometimes cause the
filewatcher to be triggered twice (once when opening and truncating the
file, once when the file is closed and writing is done). Add a helper
function to do the writing on a separated file and then use os.Rename
for an atomic move instead.

### CODE REVIEW

File: filewatcher/filewatcher_test.go [hunk 1]
```diff
@@ -79,16 +98,7 @@ func TestFileWatcher(t *testing.T) {
...
-		f, err := os.OpenFile(path, os.O_WRONLY|os.O_TRUNC, 0644)
-		if err != nil {
-			t.Fatal(err)
-		}
-		defer f.Close()
-		if _, err := f.Write(payload2); err != nil {
-			t.Fatal(err)
-		}
-	}()
+	overwriteFile(t, path, bytes.NewReader(payload2))
```
Comment (kylelemons) [inline filewatcher/filewatcher_test.go:101]: 🔕 you use bytes.NewReader for both, so you might as well just pass the contents as `[]byte` and use `os.WriteFile`

**Review by kylelemons — APPROVED** [review]

File: filewatcher/filewatcher_test.go (hunk 1 above)
Comment (fishy) [reply filewatcher/filewatcher_test.go:101]: Looking at the source code of `os.WriteFile`, it tries to do a single `Write`, which I don't really like and it's more likely to fail if we are trying to write large content (although we don't write large content here). While `io.Copy` handles those cases much more thoroughly.

File: filewatcher/filewatcher_test.go [hunk 2]
```diff
@@ -20,6 +20,25 @@ func parser(f io.Reader) (interface{}, error) {
 	return io.ReadAll(f)
 }
 
+func overwriteFile(tb testing.TB, path string, content io.Reader) {
```
Comment (pacejackson) [inline filewatcher/filewatcher_test.go:23]: 🔕  might be good to add a comment to this explaining why we have it, at least for future readers learning?

**Review by pacejackson — APPROVED** [review]

File: filewatcher/filewatcher_test.go (hunk 2 above)
Comment (fishy) [reply filewatcher/filewatcher_test.go:31]: Done.

**Comment by fishy** [comment]
💇 added comment to explain `overwriteFile`.

File: filewatcher/filewatcher_test.go (hunk 1 above)
Comment (kylelemons) [reply filewatcher/filewatcher_test.go:109]: You can refactor `overwriteFile` in the future _if_ you end up with a reader.  YAGNI / KISS / premature optimization / etc.

**Review by kylelemons — APPROVED** [review]
(to appease harold)

File: filewatcher/filewatcher_test.go (hunk 1 above)
Comment (fishy) [reply filewatcher/filewatcher_test.go]: Changed to use `os.WriteFile`. Looks like `os.File.Write` itself implemented big write handling, at least for linux.

**Pull request merged**